core/utils/exact_match_table.h

```diff
@@ -0,0 +1,440 @@
+    return entry ? entry->second : default_value;
+  }
+
+  // Find entries for `n` `keys` in the table and store their values in in
+  // `vals`.  Keys without entries will have their corresponding entires in
+  // `vals` set to `default_value`.
+  void Find(const ExactMatchKey *keys, T *vals, size_t n,
+            T default_value) const {
+    const auto &table = table_;
+    for (size_t i = 0; i < n; i++) {
+      const auto *entry =
+          table.Find(keys[i], ExactMatchKeyHash(total_key_size_),
+                     ExactMatchKeyEq(total_key_size_));
+      vals[i] = entry ? entry->second : default_value;
+    }
+  }
+
+  uint32_t total_key_size() const { return total_key_size_; }
+
+  // Set the `idx`th field of this table to one at offset `offset` bytes into a
+  // buffer with length `size` and mask `mask`.
+  // Returns 0 on success, non-zero errno on failure.
+  Error AddField(int offset, int size, uint64_t mask, int idx) {
+    ExactMatchField f = {
+        .mask = mask, .attr_id = 0, .offset = offset, .pos = 0, .size = size};
+    return __AddField(f, "", idx, nullptr);
+  }
+
+  // Set the `idx`th field of this table to one at the offset of the
+  // `mt_attr_name` metadata field as seen by module `m`, with length `size` and
+  // mask `mask`.
+  // Returns 0 on success, non-zero errno on failure.
+  Error AddField(Module *m, const std::string &mt_attr_name, int size,
+                 uint64_t mask, int idx) {
+    ExactMatchField f = {
+        .mask = mask, .attr_id = 0, .offset = 0, .pos = 0, .size = size};
+    return __AddField(f, mt_attr_name, idx, m);
+  }
+
+  size_t num_fields() const { return num_fields_; }
+
+  // Returns the ith field.
+  const ExactMatchField &get_field(size_t i) const { return fields_[i]; }
+
+ private:
+  Error MakeError(int code, const std::string &msg) {
+    return std::make_pair(code, msg);
+  }
+
+  // Turn a rule into a key.
+  // Returns 0 on success, non-zero errno on failure.
+  Error gather_key(const ExactMatchRuleFields &fields, ExactMatchKey *key) {
+    if (fields.size() != num_fields_) {
+      return MakeError(EINVAL, Format("rule should have %zu fields (has %zu)",
+                                      num_fields_, fields.size()));
+    }
+
+    memset(key, 0, sizeof(*key));
+
+    for (size_t i = 0; i < fields.size(); i++) {
+      int field_size = fields_[i].size;
+      int field_pos = fields_[i].pos;
+
+      const std::vector<uint8_t> &f_obj = fields[i];
+
+      if (static_cast<size_t>(field_size) != f_obj.size()) {
+        return MakeError(
+            EINVAL, Format("rule field %zu should have size %d (has %zu)", i,
+                           field_size, f_obj.size()));
+      }
+
+      memcpy(reinterpret_cast<uint8_t *>(key) + field_pos, f_obj.data(),
+             field_size);
+    }
+
+    return MakeError(0, std::string(""));
+  }
+
+  // Helper for public MakeKey functions
+  void __MakeKeys(ExactMatchKey *keys, const void **bufs, size_t n) const {
```

Comment:
Single `_` should be used here, since we are not compiler or low level library. Or `DoMakeKeys` naming convention can be used.